### ISSUE TITLE
Unsuccessful permutation with pandas objects in scoring_data

### ISSUE THREAD

**Issue by chiemvs** [gelijergensen/PermutationImportance#84]
I noticed equal single-pass importance scores for input variables that should not have equal information value. This happened regardless of evaluation_fn, nbootstrap, etc. The problem seems to present itself when passing a scoring_data tuple of `(pd.DataFrame, pd.DataFrame)` and not when passing `(np.ndarray, np.ndarray)`. I think that the permutation is negated by `pd.concat` in `PermutationImportance.utils.make_data_from_columns`, as it re-aligns the data by its index. Here is a minimum working example where input variable A should have all importance and variable (unrelated noise) should have none.
```python
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error 
from PermutationImportance import sklearn_permutation_importance

length = 1000
do_datetime = True # Whether the pandas index is datetime or not does not make a difference, only the fact that pandas aligns all indices
if do_datetime:
    index = pd.Index(pd.date_range(start = '2000-01-01', periods = length))
else:
    index = pd.RangeIndex(length)
X = pd.DataFrame({'A': np.random.normal(size = length), 'B':np.random.normal(size = length)}, index = index) # y will be a linear combination of A, B is unrelated noise
y = X.loc[:,['A']] * 2 + 3 

X_train = X.iloc[:int(0.8*length),:]
y_train = y.iloc[:int(0.8*length),:]
X_val = X.iloc[int(0.8*length):,:]
y_val = y.iloc[int(0.8*length):,:]

model = LinearRegression()
model.fit(X = X_train, y = y_train)

"""
A and B have equal singlepass importance when using pandas frames. Both very low mean_absolute_error. As if permutation did not happen.
"""
result_pd = sklearn_permutation_importance(model = model, scoring_data = (X_val, y_val), evaluation_fn = mean_absolute_error, scoring_strategy = 'argmax_of_mean', njobs = 1, nbootstrap = 1)

print('pandas:')
print(f'singlepass: {result_pd.retrieve_singlepass()}')
print(f'multipass: {result_pd.retrieve_multipass()}')

"""
A and B have unequal singlepass importance when using numpy. As expected permuting A results in very high mean_absolute_error, not for B, which is unimportant
"""
result_np = sklearn_permutation_importance(model = model, scoring_data = (X_val.values, y_val.values), variable_names = X_val.columns, evaluation_fn = mean_absolute_error, scoring_strategy = 'argmax_of_mean', njobs = 1, nbootstrap = 1)

print('numpy:')
print(f'singlepass: {result_np.retrieve_singlepass()}')
print(f'multipass: {result_np.retrieve_multipass()}')

"""
I believe the problem lies in the alignment of pandas objects
by pd.concat. Called in PermutationImportance.utils.make_data_from_columns
"""
indices = np.random.permutation(len(X_val))
A_permuted = X_val.iloc[indices,X_val.columns.tolist().index('A')]
print('A == A_permuted?', X_val['A'].equals(A_permuted))
X_val_permuted = pd.concat([A_permuted,X_val['B']], axis = 1)
print('X_val == X_val_permuted?', X_val.equals(X_val_permuted))

```

```
pip3 list 
Package               Version
--------------------- -------
joblib                0.17.0 
numpy                 1.19.2 
pandas                1.1.2  
PermutationImportance 1.2.1.6
pip                   20.0.2 
pkg-resources         0.0.0  
python-dateutil       2.8.1  
pytz                  2020.1 
scikit-learn          0.23.2 
scipy                 1.5.2  
setuptools            44.0.0 
six                   1.15.0 
threadpoolctl         2.1.0 
```

**Comment by gelijergensen** [comment]
Good catch on finding this bug and thank you very much for the MWE! You were indeed correct that the problem was due to the alignment of pandas objects in `pd.concat`. The fix was simply to do an inner join instead of an outer join, e.g.
```
X_val_permuted = pd.concat([A_permuted,X_val['B']], axis = 1, join="inner")
print('X_val == X_val_permuted?', X_val.equals(X_val_permuted))
```
I have uploaded the fix both here and to PyPI, so you should see everything working correctly with version 1.2.1.7

**Comment by chiemvs** [comment]
My pleasure! Thank you for your efforts on this package. I am however not a 100% sure your fix works. You get a False because the 'inner' join forces the permuted order of A also on B, making the index of the total concatenated object take the permuted order. So basically by concatenating you have now shuffled both. Sorting the index reveals that values are still matched by index:
```python
X_val_permuted_inner = pd.concat([A_permuted,X_val['B']], axis = 1, join = 'inner')
print('X_val == X_val_permuted_inner?', X_val.equals(X_val_permuted_inner))
print('X_val == X_val_permuted_inner_sort?', X_val.equals(X_val_permuted_inner.sort_index()))
```
Am I correct that your `make_data_from_columns` has to handle a collection of permuted and non-permuted columns in any order or quantity? In that case a solution like re-building the total frame with e.g. `pd.DataFrame(data = [column1.values, column2.values], columns = variable_names,  index = column1.index)` will not work, as there is no guarantee that you'll pass a meaningfully ordered first_column index for the scoring_inputs. But also the opposite option of handling everything as numpy arrays and feeding those to the fitted models, might mess-up someone's custom (non-sklearn) scoring_fn dependent on pandas input. Perhaps it is an option to keep track of the originally supplied index when a user wants to work with pandas?

**Comment by gelijergensen** [comment]
Okay, I'm pretty sure I've got it this time. We have to strip off the index before concatenating and then we can put it back on afterwards. Thus, the function becomes:
```python
def make_data_from_columns(columns_list, index=None):
   # ...
    if isinstance(columns_list[0], pd.DataFrame) or isinstance(columns_list[0], pd.Series):
        df = pd.concat([c.reset_index(drop=True) for c in columns_list], axis=1)
        if index is not None:
            return df.set_index(index)
        else:
            return df
    # ...
```
So we now have 
```python
correctly_permuted = make_data_from_columns([A_permuted, X_val["B"]], index=X_val.index)
````
which gives the expected result where only column A is permuted
```
X_val:
                   A         B
2000-02-10  0.753740 -0.416511
2000-02-11 -0.785925 -0.570386
2000-02-12  1.150884  1.291794
2000-02-13 -0.375025  1.012964
2000-02-14 -0.656985  0.233263
2000-02-15  0.548382 -0.665439
2000-02-16 -0.392609  0.934708
2000-02-17  0.216557 -0.244008
2000-02-18 -0.760954 -0.654842
2000-02-19  1.566288  0.221165
A_permuted:
2000-02-16   -0.392609
2000-02-18   -0.760954
2000-02-14   -0.656985
2000-02-12    1.150884
2000-02-10    0.753740
2000-02-19    1.566288
2000-02-17    0.216557
2000-02-11   -0.785925
2000-02-13   -0.375025
2000-02-15    0.548382
Name: A, dtype: float64
correctly_permuted:
                   A         B
2000-02-10 -0.392609 -0.416511
2000-02-11 -0.760954 -0.570386
2000-02-12 -0.656985  1.291794
2000-02-13  1.150884  1.012964
2000-02-14  0.753740  0.233263
2000-02-15  1.566288 -0.665439
2000-02-16  0.216557  0.934708
2000-02-17 -0.785925 -0.244008
2000-02-18 -0.375025 -0.654842
2000-02-19  0.548382  0.221165
X_val == correctly_permuted? False
X_val == correctly_permuted_sort? False
```

**Comment by gelijergensen** [comment]
And yes, we can just keep track of the originally supplied index. I've also corrected the `PermutationImportanceSelectionStrategy` to handle everything, so no change is required by the end user

**Comment by chiemvs** [comment]
Looks good! Many thanks. I hope the bug didn't affect any of your past work. For me it was the first time using the package, and immediately incomprehensible results puzzled me. But it all looks better now! Cheers

**Comment by gelijergensen** [comment]
Glad that fixed it! And fortunately I had done all my work in the past with numpy arrays, which is probably why this bug slipped by me originally
